### ISSUE TITLE
False positive: array with optional keys does not accept generic array

### ISSUE THREAD

**Issue by MidnightDesign** [vimeo/psalm#6074]
```php
/**
 * @param array{label?: string} $data
 * @return string|null
 */
function extractLabel(array $data): ?string {
    return $data['label'] ?? null;
}

/** @var array<string, string> */
$data = ['foo' => 'bar'];
extractLabel($data);
```
https://psalm.dev/r/a1ba2a76d8

### Expected
No errors; the argument of type `array{label?: string}` accepts a `array<string, string>`.

### Actual
Psalm reports
> ERROR: ArgumentTypeCoercion - 13:14 - Argument 1 of extractLabel expects array{label?: string}, parent type array<string, string> provided

[The equivalent TypeScript snippet](https://www.typescriptlang.org/play?#code/MYewdgzgLgBApgDygJwIbCgGVQIzgGxgF4YAKAE1SlQC4YBvfXAgfjumQEswBzAXwCU7FNx4AfMAFd8hIgD4GAKBgqYyOFEnIwMStQB0TPIRYsYUmQG5Ffa4tCRYe1MQYAzECDoByHKmTefDCoEDAASnCgyOQAPByiADQw8bxy1ogo6FjM+BRUqAKWQA) (No errors)

**Comment by orklah** [comment]
I must say I rather agree with Psalm on this one.

I wonder why Typescript does not report an issue for something like that.
